Added evolution file to apply foreign key constraints to database

Jetpackjules opened the pull request:
Resolves #3574

The issue was that missing foreign key constraints would allow for certain incorrect insertions or deletions (Like null), and  "removing data wasn't cascading to all the tables you'd expect." I've added a DB conf evolution that should manually add all the proper foreign key constraints and fix the issue.

##### BEFORE:
![image](https://github.com/user-attachments/assets/6984e9aa-48ff-4b40-aac6-035f5b3e0dc9)

##### AFTER:
![image](https://github.com/user-attachments/assets/669b9336-06b7-4817-acf0-dcf4f57e7bc2)


##### Testing instructions
To check if the fix worked, check foreign key constraints in the relevant database with:
Run these lines in a separate WSL terminal in the root directory of the project:
> docker exec -it projectsidewalk-db bash

> psql -U postgres -d sidewalk

>SET search_path TO DATABASE_USER_HERE;

>SELECT conname AS constraint_name, conrelid::regclass AS table_name, pg_get_constraintdef(oid) AS definition
>FROM   pg_constraint
>WHERE  contype = 'f'
>AND    connamespace = 'sidewalk_walla_walla'::regnamespace;

(The above should output a table listing all foreign key constraints)



##### Things to check before submitting the PR <!-- if something doesn't apply, just check the box or remove the line -->
<!-- You can check the box by replacing the space with an "x". So instead of "- [ ]" it would be "- [x]" -->

- [X] I've included before/after screenshots above.
- [X] I've asked for and included translations for any user facing text that was added or modified.
- [X] I've updated any logging. Clicks, keyboard presses, and other user interactions should be logged. If you're not sure how (or if you need to update the logging), ask Mikey. Then make sure the documentation on [this wiki page](https://github.com/ProjectSidewalk/SidewalkWebpage/wiki/Descriptions-of-Logged-Events) is up to date for the logs you added/updated.
- [X] I've tested on mobile (only needed for the validation page).

Jetpackjules committed:
Added evolution to apply foreign key constraints to database, addressing issue 3574

misaugstad reviewed (changes requested):
@Jetpackjules these are small things to clean up, but I'll throw them to you now so that you can change them before I get a chance to fully review/test this!

1. I don't see the comments as adding anything, really. It's basically as verbose as the SQL itself, and the SQL is pretty self explanatory, so I think that we can just remove these comments.
2. Let's reduce the length of the file slightly by combining the last two lines of each:
    ```
    ALTER TABLE audit_task
        ADD CONSTRAINT fk_audit_task_amt_assignment
            FOREIGN KEY (amt_assignment_id) REFERENCES amt_assignment(amt_assignment_id);
    ```
3. And for the Down section, let's just make each of those one line!
4. For the Downs section, technically they should be in reverse order compared to the Ups... If the Downs are "undoing" the Ups, then we go in reverse order! Doesn't have any practical implications for this file, but we should stick to it because it _sometimes_ matters!

Jetpackjules committed:
Update and rename 251.sql to 252.sql

I changed the evolution number to fix the conflict and did the simple requested reformatting.

Jetpackjules commented:
Thanks for the preliminary feedback, I've implemented the changes!

misaugstad commented on conf/evolutions/default/252.sql:
```diff
@@ -0,0 +1,129 @@
+# --- !Ups
+
+ALTER TABLE audit_task
+    ADD CONSTRAINT fk_audit_task_amt_assignment
```
Instead of naming the foreign keys explicitly, let's just use the names that they give to the constraints by default. This is how we've done it elsewhere, so we should continue with that so that naming conventions aren't changing! I believe that the lines should look more like:
```
ALTER TABLE audit_task
    ADD FOREIGN KEY (amt_assignment_id) REFERENCES amt_assignment(amt_assignment_id);
```

Then it will default to naming the keys `<table-name>_<column-name>_fkey`. So for the first one, it would be `audit_task_amt_assignment_id_fkey`. You'll still need to know what they'll be called for the downs!

misaugstad commented on conf/evolutions/default/252.sql:
```diff
@@ -0,0 +1,129 @@
+
+# --- !Downs
+
+DROP CONSTRAINT IF EXISTS fk_validation_task_interaction_mission;
```
Did the Downs work for you? I'm pretty sure that you need to specify the name of the table? So it would be:
```
ALTER TABLE validation_task_interaction DROP CONSTRAINT IF EXISTS fk_validation_task_interaction_mission;
```

misaugstad reviewed (changes requested):
Another couple things to change before I test!

Jetpackjules committed:
Update 252.sql

Used default constraint names and fixed downs by specifying table

Jetpackjules committed:
Rename 252.sql to 253.sql

jonfroehlich commented:
Checking in on this one, @Jetpackjules. After you addressed the first round of review feedback, @misaugstad left a second review ("Another couple things to change before I test!") but the specific items weren't spelled out in the thread. The branch now has merge conflicts against `develop`.

Can you ping @misaugstad to find out what those remaining items were, then address them and rebase? Adding the missing FK constraints is still on the backlog — this is worth finishing.

Jetpackjules commented:
Hey Prof Froehlich, thanks for checking in on this. From what I remember, I had already addressed Mikey’s review feedback in the later commits: switching back to the default FK constraint names, fixing the Downs syntax, and renaming the evolution file from 252.sql to 253.sql.

Unfortunately, since this PR is from a while ago (I no longer have the repo locally) and I’m now working full-time plus another lab, I don’t think I have the bandwidth to properly pick it back up (set up the repo, etc), rebase it, and test/debug any conflicts. If this is still important, it may make more sense for someone else to take it over from here. I’m happy to provide context on the original changes if that would help!

jonfroehlich commented:
Thanks for all the work here, @Jetpackjules — and no worries at all about bandwidth, that's completely understandable.

Digging into the state of this, the fair read is that it stalled on *our* side. You addressed both rounds of @misaugstad's feedback back in Nov 2024 (default constraint names, table-qualified `DROP`s in the Downs, renumbered file); the follow-up "before I test" review just never got its test pass, and then it sat.

Two things have changed underneath it since, so it's unfortunately not a clean rebase:

1. **The auth/user tables moved into a shared `sidewalk_login` schema.** `user_role`, `user_login_info`, and `user_password_info` no longer live in the per-city schema, and `sidewalk_user`/`role`/`login_info` moved with them. That invalidates 8 of the 24 `ALTER`s here — 4 target tables that aren't in the city schema anymore (the evolution would error out), and 4 reference the moved tables cross-schema. The ~16 non-auth FKs are still valid and still missing (confirmed against the live db), so that portion is still good.
2. **These are plain FKs (no `ON DELETE CASCADE`).** The original motivation in #3574 was deletes not cascading for PII removal — but a bare FK *restricts* deletes rather than cascading them, so on the user-linked tables this would make PII deletion harder, not easier. Worth a deliberate decision on which of these should cascade.

Given the schema drift, and since this overlaps your ongoing backend SQL cleanup (and the cascade call is really yours), @misaugstad — I think the cleanest path is to close this PR, keep #3574 open, and fold the remaining constraints into that cleanup, reusing Jules's non-auth list as a starting point and adding the `sidewalk_login` FKs separately. I'll leave a summary on #3574 so none of this context is lost. Closing for now — sound right?

Thanks again @Jetpackjules, this was a solid start and most of it holds up.
